### PR TITLE
Feature/observe malicious patients

### PR DESCRIPTION
**Pull request by Jana211**

### CODE REVIEW

File: PatientWebApplication/HealthClinic.DAL/Model/Patient/PatientUser.cs [hunk 1]
```diff
@@ -103,6 +107,7 @@ public PatientUser() : base()
             this.parentName = parentName;
             this.exLastname = exLastname;
             this.file = file;
+            isBlocked = false;
```
Comment (velicko-tam) [inline PatientWebApplication/HealthClinic.DAL/Model/Patient/PatientUser.cs:110]: Change isBlocked = false into this.isBlocked = false

File: PatientWebApplication/HealthClinic.DAL/Model/Patient/PatientUser.cs [hunk 2]
```diff
@@ -123,6 +128,7 @@ public PatientUser() : base()
             this.parentName = parentName;
             this.exLastname = "";
             this.file = file;
+            isBlocked = false;
```
Comment (velicko-tam) [inline PatientWebApplication/HealthClinic.DAL/Model/Patient/PatientUser.cs:131]: Change isBlocked = false into this.isBlocked = false

File: PatientWebApplication/HealthClinic.DAL/Model/Patient/PatientUser.cs [hunk 3]
```diff
@@ -80,6 +83,7 @@ public PatientUser() : base()
...
```
Comment (velicko-tam) [inline PatientWebApplication/HealthClinic.DAL/Model/Patient/PatientUser.cs:86]: Change isBlocked = false into this.isBlocked = false

File: PatientWebApplication/HealthClinic.DAL/Model/Patient/PatientUser.cs [hunk 4]
```diff
@@ -60,6 +62,7 @@ public PatientUser() : base()
             this.parentName = parentName;
             this.exLastname = exLastname;
             this.file = file;
+            isBlocked = false;
```
Comment (velicko-tam) [inline PatientWebApplication/HealthClinic.DAL/Model/Patient/PatientUser.cs:65]: Change isBlocked = false into this.isBlocked = false

File: PatientWebApplication/PatientWebApplicationTests/BlockPatientTests.cs [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+            patient.ShouldNotBeNull();
+        }
+
+        private static IPatientsRepository CreateStubRepository()
+        {
+            var stubRepository = new Mock<IPatientsRepository>();
+
+            var patients = new List<PatientUser>();
+
+            PatientUser patient1 = new PatientUser(1, "Pera2", "Peric", "Male", "1234", "2/2/2020", "123", "212313", "Alergija", "Grad", false, "email", "pass", false, "Grad2", "Roditelj", null);
```
Comment (velicko-tam) [inline PatientWebApplication/PatientWebApplicationTests/BlockPatientTests.cs:46]: Change PatientUser patient1 name

File: PatientWebApplication/PatientWebApplicationTests/BlockPatientTests.cs [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+            return stubRepository.Object;
+        }
+
+        private static IPatientsRepository CreateStubRepositoryMalicious()
+        {
+            var stubRepository = new Mock<IPatientsRepository>();
+
+            var patients = new List<PatientUser>();
+
+            PatientUser patient1 = new PatientUser(1, "Pera2", "Peric", "Male", "1234", "2/2/2020", "123", "212313", "Alergija", "Grad", false, "email", "pass", false, "Grad2", "Roditelj", null);
```
Comment (velicko-tam) [inline PatientWebApplication/PatientWebApplicationTests/BlockPatientTests.cs:68]: Change PatientUser patient1 name

File: PatientWebApplication/HealthClinic.DAL/Service/PatientService.cs [hunk 7]
```diff
@@ -84,13 +87,63 @@ public string ImageToSave(string path, FileModel file)
...
+            foreach (int key in dict.Keys)
+            {
+                if (dict[key] >= 3)
+                    maliciousPatients.Add(GetOne(key));
+            }
+            return maliciousPatients;
+        }
+        private Dictionary<int, int> GetCanceledAppointmentsInLastMonth(List<DoctorAppointment> appointments) {
+            Dictionary<int, int> dict = new Dictionary<int, int>();
+            foreach (DoctorAppointment appointment in appointments)
```
Comment (velicko-tam) [inline PatientWebApplication/HealthClinic.DAL/Service/PatientService.cs:131]: Refactor method GetCanceledAppointmentsInLastMonth

**Review by velicko-tam — CHANGES_REQUESTED** [review]
PR is not approved. Marked parts need to be fixed.

File: PatientWebApplication/PatientWebApplicationTests/BlockPatientTests.cs [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+            patient.ShouldNotBeNull();
+        }
+
+        private static IPatientsRepository CreateStubRepository()
+        {
+            var stubRepository = new Mock<IPatientsRepository>();
+
+            var patients = new List<PatientUser>();
+
+            PatientUser BlockedPatient = new PatientUser(1, "Pera2", "Peric", "Male", "1234", "2/2/2020", "123", "212313", "Alergija", "Grad", false, "email", "pass", false, "Grad2", "Roditelj", null);
```
Comment (velicko-tam) [inline PatientWebApplication/PatientWebApplicationTests/BlockPatientTests.cs:46]: BlockedPatient shuld not start with a capitol letter.

File: PatientWebApplication/PatientWebApplicationTests/BlockPatientTests.cs [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+          
+            patients.Add(BlockedPatient);
+
+            stubRepository.Setup(m => m.FindOne(1)).Returns(patients.SingleOrDefault(pat => pat.id == 1));
+            stubRepository.Setup(m => m.BlockPatient(It.IsAny<PatientUser>())).Callback((PatientUser patient) =>
+            {
+                PatientUser pat = (patients.SingleOrDefault(pat => pat.id == 1));
```
Comment (velicko-tam) [inline PatientWebApplication/PatientWebApplicationTests/BlockPatientTests.cs:53]: Change PatientUser pat name.

File: PatientWebApplication/PatientWebApplicationTests/BlockPatientTests.cs [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+            return stubRepository.Object;
+        }
+
+        private static IPatientsRepository CreateStubRepositoryMalicious()
+        {
+            var stubRepository = new Mock<IPatientsRepository>();
+
+            var patients = new List<PatientUser>();
+
+            PatientUser MaliciousPatient = new PatientUser(1, "Pera2", "Peric", "Male", "1234", "2/2/2020", "123", "212313", "Alergija", "Grad", false, "email", "pass", false, "Grad2", "Roditelj", null);
```
Comment (velicko-tam) [inline PatientWebApplication/PatientWebApplicationTests/BlockPatientTests.cs:68]: MaliciousPatient shuld not start with a capitol letter.

**Review by velicko-tam — CHANGES_REQUESTED** [review]
PR is not approved. Marked parts need to be fixed.

**Review by velicko-tam — CHANGES_REQUESTED** [review]
PR is not approved. BlockPatientTests failed on the pipeline.

**Review by velicko-tam — APPROVED** [review]
PR is approved. Clean code principles were implemented.

File: PatientWebApplication/HealthClinic.DAL/Service/PatientService.cs [hunk 11]
```diff
@@ -84,13 +87,70 @@ public string ImageToSave(string path, FileModel file)
...
+            PatientUser patient = PatientsRepository.FindOne(patientId);
+            return (patient == null) ? null : PatientsRepository.BlockPatient(patient);
+        }
+
+        /// <summary> This method is getting list of filtered malicious<c>PatientUser</c>.</summary>
+        /// <returns> List of filtered malicious patients. </returns>
+        public List<PatientUser> GetMaliciousPatients()
+        {
+            Dictionary<int,int> dict = new Dictionary<int, int>();
+            List<PatientUser> maliciousPatients = new List<PatientUser>();
```
Comment (trki293) [inline PatientWebApplication/HealthClinic.DAL/Service/PatientService.cs:114]: An instance list that has not been used. Need to delete.

File: PatientWebApplication/HealthClinic.DAL/Service/PatientService.cs [hunk 12]
```diff
@@ -84,13 +87,70 @@ public string ImageToSave(string path, FileModel file)
...
+            List<DoctorAppointment> appointments = _regularAppointmentService.GetAll();
+            dict = GetCanceledAppointmentsInLastMonth(appointments);
+            return GetValidPatientsInLastMonth(dict);
```
Comment (trki293) [inline PatientWebApplication/HealthClinic.DAL/Service/PatientService.cs:117]: Write this in form:
"return GetValidPatientsInLastMonth(GetCanceledAppointmentsInLastMonth(_regularAppointmentService.GetAll()));"

**Review by trki293 — CHANGES_REQUESTED** [review]
Necessary changes in the code to meet the principles of the Clean code.

**Review by trki293 — APPROVED** [review]
PR is approved. The good principles of the Clean code have been followed.

**Pull request merged**